code/krepel/math/vector3.d

```diff
@@ -0,0 +1,222 @@
+module krepel.math.vector3;
+
+import krepel.math.math;
+
+float Dot(Vector3 lhs, Vector3 rhs)
+{
+  float result = lhs | rhs;
+  return result;
+}
+
+Vector3 Mul(Vector3 lhs, Vector3 rhs)
+{
+  return lhs * rhs;
+}
+
+Vector3 Cross(Vector3 lhs, Vector3 rhs)
+{
+  return lhs ^ rhs;
+}
+
+float LengthSquared(Vector3 vec)
+{
+  return vec | vec;
+}
+
+float Length(Vector3 vec)
+{
+  return Sqrt(vec.LengthSquared());
+}
+
+struct Vector3
+{
+  union
+  {
+    struct{
+      float X, Y, Z;
+    }
+    float[3] Data;
+  }
+  this(float Value)
+  {
+    X = Value;
+    Y = Value;
+    Z = Value;
+  }
+
+  this(float X, float Y, float Z)
+  {
+    this.X = X;
+    this.Y = Y;
+    this.Z = Z;
+  }
+
+  // Dot product
+  float opBinary(string op:"|")(Vector3 rhs)
+  {
+    return
+      X * rhs.X +
+      Y * rhs.Y +
+      Z * rhs.Z;
+  }
+
+  Vector3 opBinary(string op)(Vector3 rhs)
+  {
+    // Addition, subtraction, component wise multiplication
+    static if(op == "+" || op == "-" || op == "*")
+    {
+      return mixin("Vector3("~
+        "X" ~ op ~ "rhs.X,"
+        "Y" ~ op ~ "rhs.Y,"
+        "Z" ~ op ~ "rhs.Z)");
+    }
+    // Cross Product
+    else if(op == "^")
+    {
+      return Vector3(
+        (Y * rhs.Z) - (Z * rhs.Y),
+        (Z * rhs.X) - (X * rhs.Z),
+        (X * rhs.Y) - (Y * rhs.X)
+      );
+    }
+    else
+    {
+      assert(false, "Operator " ~ op ~ " not implemented.");
+    }
+  }
+
+  Vector3 opBinary(string op)(float rhs)
+  {
+    // Vector scaling
+    static if(op == "/" || op == "*")
+    {
+      return mixin("Vector3("~
+        "X" ~ op ~ "rhs,"
+        "Y" ~ op ~ "rhs,"
+        "Z" ~ op ~ "rhs)");
+    }
+    else
+    {
+      assert(false, "Operator " ~ op ~ " not implemented.");
+    }
+  }
+
+  Vector3 opBinaryRight(string op)(float rhs)
+  {
+    // Vector scaling
+    static if(op == "*")
+    {
+      return mixin("Vector3("~
+        "X" ~ op ~ "rhs,"
+        "Y" ~ op ~ "rhs,"
+        "Z" ~ op ~ "rhs)");
+    }
+    else
+    {
+      assert(false, "Operator " ~ op ~ " not implemented.");
+    }
+  }
+
+  static Vector3 Forward = Vector3(1,0,0);

```

Comment:
These should be `__gshared immutable Foo = Vector3(...);`.

(`immutable` and `const` act as if you stated `auto`, so there's no need to state the type twice.)